### PR TITLE
Orders placing

### PR DESCRIPTION
**Pull request by BohdanZorii**

### CODE REVIEW

**Comment by BohdanZorii** [comment]
ongoing issue fix conversation in jv_chat_1

File: src/main/java/mate/zorii/bookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Create a new order",
+            description = "Creates a new order for the currently authenticated user.")
+    public List<OrderResponseDto> placeOrder(@RequestBody @Valid CreateOrderRequestDto requestDto,
+                                             Authentication auth) {
+        Long authenticatedUserId = authService.getAuthenticatedUserId(auth);
+        List<OrderResponseDto> orders = orderService.placeOrder(requestDto, authenticatedUserId);
+        shoppingCartService.clearCart(authenticatedUserId);
+        return orders;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/controller/OrderController.java:48]:
```suggestion
        return orderService.placeOrder(user.getId(), requestDto);
```

File: src/main/java/mate/zorii/bookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+public class OrderController {
+    private final OrderService orderService;
+    private final AuthenticationService authService;
+    private final ShoppingCartService shoppingCartService;
+
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Create a new order",
+            description = "Creates a new order for the currently authenticated user.")
+    public List<OrderResponseDto> placeOrder(@RequestBody @Valid CreateOrderRequestDto requestDto,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/controller/OrderController.java:43]:
```suggestion
    public OrderDto placeOrder(@RequestBody @Valid CreateOrderRequestDto requestDto,
```

File: src/main/java/mate/zorii/bookstore/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            description = "Retrieves all orders for authenticated user with pagination support.")
+    public Page<OrderResponseDto> getAllOrders(Pageable pageable, Authentication auth) {
+        return orderService.getAllOrders(authService.getAuthenticatedUserId(auth), pageable);
+    }
+
+    @PatchMapping("/{orderId}")
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Update order status",
+            description = "Updates the status of a specific order.")
+    public List<OrderResponseDto> updateOrderStatus(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/controller/OrderController.java:63]:
```suggestion
    public UpdateOrderResponseDto updateOrderStatus(
```

File: src/main/java/mate/zorii/bookstore/service/impl/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                .map(orderMapper::toOrderItemDto)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("No order item by id %d for orderId %d", itemId, orderId)));
+    }
+
+    @Override
+    @Transactional
+    public List<OrderResponseDto> placeOrder(CreateOrderRequestDto createDto,
+                                             Long authenticatedUserId) {
+        Order order = new Order();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/OrderServiceImpl.java:73]:
```suggestion
        ShoppingCart cart = cartRepository.findByUserId(userId);
        if (cart.getCartItems().isEmpty()) {
            .....
        }
        Order order = orderMapper.cartToOrder(cart, orderDto.shippingAddress());
        cart.clearCart();
        return orderMapper.toOrderDto(orderRepository.save(order));
```

Where in OrderMapper class

```
@Mapping(target = "id", ignore = true)
@Mapping(target = "total", source = "cart.cartItems", qualifiedByName = "total")
@Mapping(target = "orderItems", source = "cart.cartItems")
Order cartToOrder(ShoppingCart cart, String shippingAddress);
```

and 

```
@Named("total")
    default BigDecimal getTotal(Set<CartItem> cartItems) {
        return cartItems.stream()
                .map(i -> i.getBook().getPrice().multiply(BigDecimal.valueOf(i.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
```

and 
```
@AfterMapping
    default void updateOrder(@MappingTarget Order order) {
        order.getOrderItems().forEach(oi -> oi.setOrder(order));
    }
```

and in ShoppingCart class
```
public void clearCart() {
        cartItems.clear();
    }
```

File: src/main/java/mate/zorii/bookstore/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", referencedColumnName = "id", nullable = false)
+    private User user;
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private Status status;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/Order.java:36]:
```suggestion
    private Status status = Status.PENDING;
```

File: src/main/java/mate/zorii/bookstore/model/Order.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/Order.java:40]:
```suggestion
    private LocalDateTime orderDate = LocalDateTime.now();
```

File: src/main/java/mate/zorii/bookstore/mapper/OrderMapper.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderMapper {
+    @Mapping(target = "userId", source = "user.id")
+    OrderResponseDto toOrderDto(Order order);
+
+    @Mapping(target = "bookId", source = "book.id")
+    OrderItemResponseDto toOrderItemDto(OrderItem orderItem);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/mapper/OrderMapper.java:18]: create OrderItemMapper for mapping logic with OrderItem

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -23,15 +26,18 @@ public class ShoppingCartServiceImpl implements ShoppingCartService {
...
     public ShoppingCartResponseDto findByUserId(Long userId) {
-        return mapper.toShoppingCartDto(shoppingCartRepository.findByUserId(userId));
+        ShoppingCart cart = shoppingCartRepository.findByUserId(userId);
+        return mapper.toShoppingCartDto(cart);
     }
 
     @Override
     @Transactional
-    public ShoppingCartResponseDto addCartItem(CartItemDto cartItemDto, Long userId) {
+    public void addCartItem(CartItemDto cartItemDto, Long userId) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:40]: revert please

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -46,7 +52,9 @@ public ShoppingCartResponseDto addCartItem(CartItemDto cartItemDto, Long userId)
                 }).orElse(mapper.toModel(cartItemDto, shoppingCart));
 
         cartItemRepository.save(cartItem);
-        return mapper.toShoppingCartDto(shoppingCart);
+        //ensure fresh data on the next select
+        entityManager.flush();
+        entityManager.clear();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:57]: what is it?

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -83,4 +91,16 @@ public void createShoppingCart(User user) {
...
     }
+
+    @Override
+    public Set<CartItem> getCartItemsByUserId(Long userId) {
+        return shoppingCartRepository.findByUserId(userId).getCartItems();
+    }
+
+    @Override
+    @Transactional
+    public void clearCart(Long userId) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:102]: remove

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -83,4 +91,16 @@ public void createShoppingCart(User user) {
         shoppingCart.setUser(user);
         shoppingCartRepository.save(shoppingCart);
     }
+
+    @Override
+    public Set<CartItem> getCartItemsByUserId(Long userId) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:96]: remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
See comments

**Comment by BohdanZorii** [comment]
get 'cannot persist detached entity orderItem', when trying to place an order

**Comment by BohdanZorii** [comment]
> get 'cannot persist detached entity orderItem', when trying to place an order

fixed by saving order to have id before setting orderItems

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Pull request merged**